### PR TITLE
expression, types: fix issue #3923, builtin function TimeDiff compability and bug.

### PR DESCRIPTION
**Pull request by winkyao**
See #3923 , and the last case is anather issue, will be handled in another pr.

### CODE REVIEW

File: util/types/etc.go [hunk 1]
```diff
@@ -68,6 +68,11 @@ func IsTypeTime(tp byte) bool {
 	return tp == mysql.TypeDatetime || tp == mysql.TypeDate || tp == mysql.TypeNewDate || tp == mysql.TypeTimestamp
 }
 
+// IsTemporalWithDate is corresponding to mysql function: is_temporal_with_date
```
Comment (XuHuaiyu) [inline util/types/etc.go]: just explain what this function does be enough.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -305,6 +305,32 @@ type builtinTimeDiffSig struct {
...
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
+	// Fix issue #3923, see https://github.com/pingcap/tidb/issues/3923,
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: can we put an example here to avoid going to the URL ?

File: expression/builtin_time.go [hunk 3]
```diff
@@ -305,6 +305,32 @@ type builtinTimeDiffSig struct {
...
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: can we change the definition of `arg` to `Expression`, and check `arg.GetClassType` in switch-case?

**Comment by winkyao** [comment]
@XuHuaiyu @zz-jason PTAL

File: expression/builtin_time.go (hunk 3 above)
Comment (zz-jason) [inline expression/builtin_time.go]: agree

File: util/types/time.go [hunk 4]
```diff
@@ -982,16 +982,25 @@ func ParseDuration(str string, fsp int) (Duration, error) {
...
-		d = MinTime
+	d, truncated := TruncateOverflowMySQLTime(d)
+	if truncated {
 		err = ErrTruncatedWrongVal.GenByArgs("time", origStr)
 	}
+
 	return Duration{Duration: d, Fsp: fsp}, errors.Trace(err)
 }
 
+// TruncateOverflowMySQLTime truncate d when it overflow, and return truncated duration.
```
Comment (XuHuaiyu) [inline util/types/time.go]: s/ overflow/ overflows

File: expression/builtin_time.go [hunk 5]
```diff
@@ -315,16 +343,31 @@ func (b *builtinTimeDiffSig) eval(row []types.Datum) (d types.Datum, err error)
 	if args[0].IsNull() || args[1].IsNull() {
 		return
 	}
+
 	sc := b.ctx.GetSessionVars().StmtCtx
-	t1, err := convertDatumToTime(sc, args[0])
+	fsp := math.Max(float64(args[0].Frac()), float64(args[1].Frac()))
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add int(xxx) here.
so we do not need to convert the type in line 349 and line 353

**Comment by winkyao** [comment]
@zz-jason @XuHuaiyu PTAL

File: util/types/time.go [hunk 6]
```diff
@@ -982,16 +982,25 @@ func ParseDuration(str string, fsp int) (Duration, error) {
...
-		d = MinTime
+	d, truncated := TruncateOverflowMySQLTime(d)
+	if truncated {
 		err = ErrTruncatedWrongVal.GenByArgs("time", origStr)
 	}
+
 	return Duration{Duration: d, Fsp: fsp}, errors.Trace(err)
 }
 
+// TruncateOverflowMySQLTime truncate d when it overflows, and return truncated duration.
```
Comment (XuHuaiyu) [inline util/types/time.go]: s/ truncate/ truncates

File: expression/builtin_time.go [hunk 7]
```diff
@@ -305,6 +305,34 @@ type builtinTimeDiffSig struct {
...
+	case types.KindString, types.KindBytes:
+		strArg := arg.GetString()
+		fsp = b.getStrFsp(strArg, fsp)
+		t, err = types.StrToDuration(sc, strArg, fsp)
+	case types.KindMysqlDuration:
+		t, err = arg.GetMysqlDuration().ConvertToTime(mysql.TypeDuration)
+	default:
+		t, err = convertDatumToTime(sc, arg)
+	}
+	return
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: better return errors.Trace(err)

File: expression/builtin_time.go [hunk 8]
```diff
@@ -305,6 +305,34 @@ type builtinTimeDiffSig struct {
...
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
+	// Fix issue #3923, see https://github.com/pingcap/tidb/issues/3923,
+	// TIMEDIFF() returns expr1 − expr2 expressed as a time value. expr1 and expr2 are time or date-and-time expressions, but both must be of the same type.
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/ time/ duration

File: expression/builtin_time.go [hunk 9]
```diff
@@ -305,6 +305,34 @@ type builtinTimeDiffSig struct {
...
+	// if expr is a string, MySQL will try to convert it to mysql time first(in TiDB,
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: If expr is a string, we first try to convert it to \`Duration\`,
if it failed to convert, we then try to convert it to \`Datetime\`.

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

File: util/types/convert.go [hunk 10]
```diff
@@ -164,6 +165,38 @@ func StrToUint(sc *variable.StatementContext, str string) (uint64, error) {
 	return uVal, errors.Trace(err)
 }
 
+// StrToDateTime convert str to MySQL DateTime.
+func StrToDateTime(str string, fsp int) (Time, error) {
+	return ParseTime(str, mysql.TypeDatetime, fsp)
+}
+
+// StrToDuration convert str to MySQL Time.
```
Comment (winkyao) [inline util/types/convert.go]: yes, StrToDuration is ported from MySQL, but StrToDateTime is not, StrToDateTime reuse TiDB previous implement.

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @zz-jason

**Pull request merged**